File: LeetCode/1499. Max Value of Equation/solution.py
class Solution:
#     def findMaxValueOfEquation(self, points: List[List[int]], k: int) -> int:
#         j, n = 1, len(points)
#         q = deque()
#         res = -float('inf')
#         for i in range(n-1):
#             j = max(j,i+1)
#             if q and i == q[0]:
#                 q.popleft()
#             while j < n and points[j][0]-points[i][0] <= k:
#                 while q and points[q[-1]][-1] + points[q[-1]][0] < points[j][-1] + points[j][0]:
#                     q.pop()
#                 q.append(j)
#                 j += 1
            
#             if q:
#                 res = max(res, points[i][-1]-points[i][0]+points[q[0]][-1]+points[q[0]][0])
#             # print(i,q,res)
#         return res


    def findMaxValueOfEquation(self, A, k):
        q = collections.deque()
        res = -float('inf')
        for x, y in A:
            while q and q[0][1] < x - k:
                q.popleft()
            if q: res = max(res, q[0][0] + y + x)
            while q and q[-1][0] <= y - x:
                q.pop()
            q.append([y - x, x])
        return res

File: LeetCode/1499. Max Value of Equation/solution.cpp
using pp = pair<int,int>;
const int inf = 0x3f3f3f3f;
const int maxn = 1e5+5;

int x[maxn], xy[maxn];

class Solution {
public:
    int findMaxValueOfEquation(vector<vector<int>>& points, int k) {
        int res = -inf; 
        int n = points.size();
        for (int left = 0, right = -1, i = 0; i < n; i++){
            int x2 = points[i][0], y2 = points[i][1];
            while (left <= right && x[left] < x2-k)
                left++;
            if (left <= right){
                // cout << x1 << " " << xy1 << " " << x2 << " " << y2 << endl;
                res = max(res, xy[left]+x2+y2);
            }
            
            while (left <= right && xy[right] <= y2-x2)
                right--;
            right++;
            x[right] = x2;
            xy[right] = y2-x2;
        }
        return res;
    }
};


class Solution {
public:

    int findMaxValueOfEquation(vector<vector<int>>& points, int k) {
        deque<pair<int, int>> dq;
        int ans = INT_MIN;
        for (auto& j : points) {
            while (!dq.empty() && j[0] - dq.front().second > k) dq.pop_front();
            if (!dq.empty()) ans = max(ans, j[0] + j[1] + dq.front().first);
            while (!dq.empty() && dq.back().first <= j[1] - j[0]) dq.pop_back();
            dq.push_back({j[1] - j[0], j[0]});
        }
        return ans;
    }
};



using pp = pair<int,int>;
const int inf = 0x3f3f3f3f;
class Solution {
public:
    int findMaxValueOfEquation(vector<vector<int>>& points, int k) {
        int res = -inf; 
        int n = points.size();
        deque<pp> q;
        for (int i = 0; i < n; i++){
            int x2 = points[i][0], y2 = points[i][1];
            while (!q.empty() && q.front().second < x2-k)
                q.pop_front();
            if (!q.empty()){
                auto [xy1, x1] = q.front();
                // cout << x1 << " " << xy1 << " " << x2 << " " << y2 << endl;
                res = max(res, xy1+x2+y2);
            }
            
            while (!q.empty() && q.back().first <= y2-x2)
                q.pop_back();
            q.push_back({y2-x2,x2});
        }
        return res;
    }
};



// using pp = pair<int,int>;
// const int inf = 0x3f3f3f3f;

// class Solution {
// public:
//     int findMaxValueOfEquation(vector<vector<int>>& points, int k) {
//         int res = -inf; 
//         int n = points.size();
//         priority_queue<pp,vector<pp>> q;
//         for (int left = 0, right = 0; left < n; left++){
//             while (right < n && points[right][0]-points[left][0] <= k){
//                 q.push({points[right][1]+points[right][0],points[right][0]});
//                 right++;
//             }
//             while (!q.empty() && q.top().second <= points[left][0])
//                 q.pop();
//             if (q.empty())
//                 continue;
//             auto [xy2,x2] = q.top(); 
//             int x1 = points[left][0], y1 = points[left][1];
//             // cout << x1 << " " << y1 << " " << xy2 << endl;
//             res = max(res, xy2+y1-x1);
//         }
//         return res;
//     }
// };